src/js.c

```diff
@@ -1850,23 +1839,22 @@ _fetch(natsMsgList *list, natsSubscription *sub, jsFetchRequest *req, bool simpl
         rply = (const char*) reqSubj;
     if ((s == NATS_OK) && req->Heartbeat)
     {
-        int64_t hbi = req->Heartbeat / 1000000;
-        sub->refs++;
-        if (jsi->hbTimer == NULL)
+        s = nats_createControlMessages(sub);
```

Comment:
best viewed with `-w` here , instead of `jsi->mhMsg` now use `sub->control` which keeps all distinct synthetic messages for the sub.